keps/sig-apps/20190326-stateful-application-api.md

```diff
@@ -0,0 +1,790 @@
+
+```yaml
+apiVersion: apps.k8s.io/v1alpha1
+kind: HookAction
+metadata:
+  name: mongo-pre-action
+spec:
+  action:
+    exec:
+      command:
+      - "mongo"
+      - "--eval"
+      - '"printjson(db.fsyncLock())"'
+---
+apiVersion: apps.k8s.io/v1alpha1
+kind: HookAction
+metadata:
+  name: mongo-post-action
+spec:
+  action:
+    exec:
+      command:
+      - "mongo"
+      - "--eval"
+      - '"printjson(db.fsyncUnlock())"'
+```
+
+To take a snapshot of a `StatefulApplication`, a user creates an `ApplicationSnapshot` resource that references the `StatefulApplication` object above in the `source` field, and expects the creation of both a set of `VolumeSnapshot` objects and an `ApplicationSnapshotContent` resource.
+
+```yaml  
+apiVersion: apps.k8s.io/v1alpha1
+kind: ApplicationSnapshot
+metadata:
+  name: mongo-snapshot
+  namespace: default
+spec:
+  snapshotClassName: mongo-class
+  snapshotContentName: mongo-snapshot-content
+  tags:
+    "mongo-version": "4.0"
+  source:
+    type: Application
+    applicationName: mongo # name of the StatefulApplication object
+  volumeSnapshotTimeoutSeconds: 300
+```
+
+#### Restoring (or Cloning) an Application from a Snapshot
+
+To restore (or clone) an application from a snapshot, a user creates a new `StatefulApplication` resource with optionally the name of the restored application (which will become the name of the `Application` resource of the restored application) and a source pointing to an `ApplicationSnapshot` resource. The user expects the creation of the Kubernetes resources (including `PersistentVolumeClaims` if any) that make up the application and an `Application` resource.
+
+```yaml
+apiVersion: apps.k8s.io/v1alpha1
+kind: StatefulApplication
+metadata:
+ name: mongo-restore
+ namespace: ns2 # target namespace of the restored application
+spec:
+ applicationName: mongo-clone # name of the Application object
+ source:
+   type: Snapshot
+   snapshotRef:
+     name: mongo-snapshot
+     namespace: ns1
+ valueSubstitutionRules:
+ - type: Name
+   oldValue: mongo
+   newValue: mongo-clone
+ - type: Label
+   key: "app.kubernetes.io/name"
+   oldValue: mongo
+   newValue: mongo-clone
+```
+
+Note that the definition above specifies two `ValueSubstitutionRules` that replaces `mongo` in names and values of label `app.kubernetes.io/name` in the application's Kubernetes resources with `mongo-clone`. Note that all appearances of label `app.kubernetes.io/name` are considered, including those in label selectors of certain resources, e.g., StatefulSets and Services.
+
+An application may be restored or cloned into a namespace that is different than the one it used to run and in which the snapshot or backup used for restoration or clone was taken. This requires support for cross-namespace transfer of `VolumeSnapshot`s, which is proposed in this [KEP](https://github.com/kubernetes/enhancements/pull/643).
+
```

Comment:
Namespace transfer KEP is moved to here: https://github.com/kubernetes/enhancements/pull/1112